Adding FC and Relu QNNPACK ops to C10 registry

supriyar opened the pull request:
Summary:
This is a preliminary change outlining the approach we plan to follow to integrate QNNPACK operators into the pytorch backend. The operators will not be made visibile to the user in the python world, so ultimately we will have a function that calls qnnpack backend based on the environment being run on.

The goal of the project is to integrate QNNPACK library with PyTorch to achieve good performance for quantized mobile models.

Differential Revision: D15806325

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+    // matrices, respectively.
+    int64_t rows_a = input_contig.size(0);
+    int64_t cols_a = 1;
+    for (size_t i = 1; i < input_contig.dim(); ++i) {
```
why are you collapsing all dimensions after the first one? I think we should do the opposite - nn.Linear supports input to be arbitrary dimensioned, but only the last dimension participates in matmul: https://pytorch.org/docs/stable/nn.html#linear

supriyar replied:
My bad, I thought it was the other way around. I'll update it to make it consistent with nn.Linear.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+      double output_scale,
+      int64_t output_zero_point) {
+
+    TORCH_CHECK(input.dim() >= 2, "Input tensor rank should be >= 2");
```
check weight to be 2d?

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
+      {rows_a, rows_b},
+      at::device(kCPU).dtype(kQUInt8),
```
more idiomatic is to just copy input.options(), but either works

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+    // Allocate output Tensor and a buffer for QNNPACK to use
+    Tensor output = at::_empty_affine_quantized(
+      {rows_a, rows_b},
+      at::device(kCPU).dtype(kQUInt8),
```
does qnnpack support only uint8? if so - let's enforce the input dtype below (by using templated data<>), if not - let's use dispatch on dtype

supriyar replied:
yes, updated to use templated data.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+        input_contig.q_scale() /* input scale */,
+        weight.q_zero_point() /* kernel zero_point */,
+        weight.q_scale() /* kernel scale */,
+        (uint8_t*)weight.data_ptr() /* kernel data */,
```
use .data<quint8_t>(), it also runs the dtype check

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+      output_zero_point);
+
+    // QNNPACK expects both weights and inputs to be uint8
+    const qnnp_status createStatus = qnnp_create_fully_connected_nc_q8(
```
how expensive is creating qnnpack op? we should consider creating a cache of those if needed

ajtulloch replied:
It's expensive, and it's also conceptually a weight-packing step - cf. https://github.com/pytorch/QNNPACK/blob/b7bacb1899e6fa3a934c1dd6128096f2e1abf071/src/fully-connected.c#L94

ajtulloch replied:
Conceptually I think this approach is wrong - like, QNNPACK is designed in a C2/TF/MXNet world of stateful operators. Essentially, we call `qnnp_create_*` in the constructor, which does packing/allocation/one-time setup. These `create_` instances can't be 'cached' because they contain state about the weights they were instantiated with. You then call `setup + run` at the `Op::Run` stage. 

Conceptually it's not really reasonable (from the perspective of achieving optimal perf) to stuff it all into the `run` call - it'll work fine but it's not right. We should either refactor QNNPACK to be more functionally oriented (which would be fairly mechanical but a decent chunk of work), or rethink this approach to match the existing calling pattern for QNNPACK + C2.

raghuramank100 replied:
I recollect this being an initial step. We should have two ops eventually:

 qnnpack_state =_qnnpack_linear_create_state(weight and other args). (This calls create)

and then have the actual op: qnnpack_linear() that performs the setup and run.

supriyar replied:
Yes, that is the plan. The plan was to submit this PR as an initial step and then work on perf optimizations by implementing caching of the create step. @dzhulgakov is that okay?

dzhulgakov replied:
Yeah, Andrew is right. Unfortunately, qnnpack itself is too static and it's easy to "break it" even in c2 bindings. For example, it bakes the quantization params for the input in op creation here: https://github.com/pytorch/pytorch/blob/28b5df1c8f8133f6bfd6326d6c83ef2578ab4518/caffe2/operators/quantized/int8_fc_op.h#L57 - so if someone calls the op second time with differently quantized input it'd break (which theoretically possible even in c2 world as we pass quantization parameters not statically but through input tensors, though it rarely happens in practice). The reason for it is that it pre-scales the weights accordingly (which is faster I suppose).

so even if we do _qnnpack_linear_create_state we'd need to pass quantization parameters for the input to it statically. It's doable as a JIT transform in most cases. I guess the best we can do as an isolated transformation is to also pass "expected" input quantization params to _qnnpack_linear_create_state and later error out if the params don't match the ones on the input tensor passed to qnnpack_linear

another problem with qnnpack is that it's not really reentrant - _setup method just modifies the "packed" structure in-place and thus the same state can't be accessed from different threads. That's pretty easy to fix though - pretty much we should change these params to be function arguments instead of fields in the object: https://github.com/pytorch/QNNPACK/blob/e95863af039a8c48e0b820a154b02f7e3d7f0690/src/fully-connected.c#L131

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_fc.cpp:
```diff
@@ -0,0 +1,98 @@
+};
+
+static auto registry = c10::RegisterOperators().op(
+    "quantized::qnnpack_fc(Tensor X, Tensor W, Tensor b, float Y_scale, int Y_zero_point) -> Tensor",
```
qnnpack_linear please

supriyar replied:
Yes, I changed to linear. I'll update the PR.

dzhulgakov commented on test/test_quantized.py:
```diff
@@ -613,6 +613,89 @@ def test_qconv(
+        np.testing.assert_equal(qY, qY_hat.int_repr())
+
+    """Tests the correctness of the quantized::qnnpack_fc op."""
+    def test_qnnpack_fc(self):
```
why don't do hypothesis testing too?

also, it would be nice to test comparison (approximately) with the float version of Linear

raghuramank100 replied:
Please ensure that 
1. qnnpack tests are identical to tests for fbgemm ops (same level of error with respect to float/reference quantized implementation)

Also, add a test to ensure that error between qnnpack and fbgemm quantized ops is small. (quantized implementations are consistent)

supriyar replied:
Done. Added hypothesis testing and also tested against FP32 results of nn.Linear

raghuramank100 reviewed (commented):
Please see comments, thanks

supriyar committed:
Adding FC and Relu QNNPACK ops to C10 registry (#22174)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22174

This is a preliminary change outlining the approach we plan to follow to integrate QNNPACK operators into the pytorch backend. The operators will not be made visible to the user in the python world, so ultimately we will have a function that calls qnnpack backend based on the environment being run on.

The goal of the project is to integrate QNNPACK library with PyTorch to achieve good performance for quantized mobile models.

Reviewed By: ljk53

Differential Revision: D15806325

fbshipit-source-id: 0f2402324ed6176872844c74f760d0333134ecbe
